### PR TITLE
Core/Gameobject: Gameobjects that are spawned by spell despawn after …

### LINKED ISSUE #12842: [3.3.5][6.x] Core/Loot/Gameobject: Tempspawned Gameobjects despawn after looting
**How it is:**
Any tempspawned chest object which is spawned with a spell from a creature (like http://www.wowhead.com/spell=46310) will despawn if **ONE single** player lootet the object.

**How it should be:**
Any tempspawned chest object which is spawned with a spell from a creature (like http://www.wowhead.com/spell=46310) should **ONLY** despawn **if the timer from the spelldescription expires** (like in the case of spell 46310 it should despawn after 1 minute)
During this time every player should be able to loot that gameobject multiple times even without a group.

Examples for quests/gos:

[Piece of Krom'zar's Banner](http://wotlk.openwow.com/item=11227)
Kaw the Mammoth Destroyer http://www.wowhead.com/spell=46310
http://www.wowhead.com/quest=10997
http://www.wowhead.com/quest=10995
http://www.wowhead.com/quest=10996
http://de.wowhead.com/quest=12713
http://www.wowhead.com/spell=18661
http://www.wowhead.com/quest=11529
http://www.wowhead.com/quest=12498
[Scythe of Antiok](http://www.wowhead.com/item=38305/scythe-of-antiok) from [Scythe of Antiok](http://www.wowhead.com/object=190354/scythe-of-antiok) for [On Ruby Wings](http://www.wowhead.com/object=190354/scythe-of-antiok).
[Drakuru's Skull](http://www.wowhead.com/item=40425/drakurus-skull) from [Drakuru's Skull](http://www.wowhead.com/object=191458/drakurus-skull).

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/18200478-db-npc-quest-grizzly-hills-giant-and-rallying-the-troops?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F1310&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by runningnak3d**
**Changes proposed**:
- To stop irritating people that beat a boss and don't get their loot because a spell summons a GO

**Target branch(es)**: 335/6x

**Issues addressed**: Fixes #12842 

**Tests performed**: builds. Tested with .add quest 12842 / .cast 46310. Didn't notice any adverse affects to statically spawned GOs.

**Known issues and TODO list**:
- [ ]  Need to test more spell summoned GOs that have no loot
- [ ] Need to test more spell summoned GOs that should only be lootable by a single player
- [x] Need to make sure that statically spawned GOs are not affected. Statically spawned GOs never get anywhere near that code -- BUT -- had to test anyway.

**NOTE** If your Pull Request is SQL only create a ticket instead

**SUGESTION** If your Pull Request contains SQL give it one imposible date, for example 9999_99_99_99_database.sql on that way it will be free on merge.

…looting. If a GO contains quest loot it should not de-spawn until the de-spawn timer from Spell.dbc expires.

Closes #12842

### CODE REVIEW

**Comment by Shauren** [comment]
Simply put, this is wrong. While it fixes objects summoned for quest like this http://www.wowhead.com/quest=11879/kaw-the-mammoth-destroyer
it also completely breaks any quest where you need to loot multiple items from summoned objects - it would allow you to complete the quest from a single object

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 1]
```diff
@@ -614,14 +614,13 @@ void GameObject::Update(uint32 diff)
...
             {
-                SetRespawnTime(0);
-                Delete();
+                UpdateObjectVisibility();
+                SetLootState(GO_READY);
                 return;
             }
-
-            SetLootState(GO_READY);
+            else Delete();
```
Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp]: please don't write else and it's statement in the same line

Comment (jackpoz) [inline src/server/game/Entities/GameObject/GameObject.cpp]: the old code returned after calling Delete(), is there any difference now not returning anymore ?

**Comment by Rushor** [comment]
maybe the summoned objects itself have some special flags in gamneobject_template which can be used to check for despawn or not despawn like for quest=11879?

File: src/server/game/Entities/GameObject/GameObject.cpp (hunk 1 above)
Comment (runningnak3d) [inline src/server/game/Entities/GameObject/GameObject.cpp]: That is a typo (deleted too much code when cleaning up), there should be a return there...

I'll fix it, if I can find a way to solve the issue Shauren brought up.

**Comment by runningnak3d** [comment]
@Rushor There doesn't need to be a flag. The duration in Spell.dbc should be something very close to 0 for those GOs that you don't want people to loot multiple times. At least that makes sense, but this is blizz we are talking about.

**Comment by runningnak3d** [comment]
@Shauren As I pointed out to @Rusher this shouldn't (if blizz wasn't bat shit crazy) break anything. There is a reason that I pointed out that some additional tests needed to be done. I assumed (which you never do) that objects like that would have a duration of 0 or something very close to it. Now that I know what I need to be looking for, I will look over those spells and see if there is anything to indicate that they should be insta-de-spawned. Lastly, as for "flat out wrong". Uh no, the code that was there was flat out wrong, and I am pretty sure that you added it :)

**Comment by Shauren** [comment]
no, duration starts from the moment the object is summoned, not looted

**Comment by runningnak3d** [comment]
@Rushor your idea about flags on the GO if they should despawn on loot is making more sense. flags&xxx then despawn when looted else despawn when duration runs out. !flags&xxx then stay spawned till duration runs out. If something can't be found from blizz, this would be a good idea for a custom flag.

**Comment by tkrokli** [comment]
You may want to push to your repo again (refresh) to restart travis-ci, now that Travis is available again. :)

File: src/server/game/Entities/GameObject/GameObject.cpp [hunk 2]
```diff
@@ -614,13 +614,18 @@ void GameObject::Update(uint32 diff)
 
             //! If this is summoned by a spell with ie. SPELL_EFFECT_SUMMON_OBJECT_WILD, with or without owner, we check respawn criteria based on spell
             //! The GetOwnerGUID() check is mostly for compatibility with hacky scripts - 99% of the time summoning should be done trough spells.
-            if ((GetSpellId() || GetOwnerGUID()) && m_respawnTime > 0)
+            //! Game objects with flags&4 should not despawn after looting so that other members of the group may loot.
+            if ((GetSpellId() || GetOwnerGUID()) && m_respawnTime > 0 && GetGOInfo()->flags & 4)
```
Comment (runningnak3d) [inline src/server/game/Entities/GameObject/GameObject.cpp]: magic number will go away if / when that flag gets renamed

**Comment by Shauren** [comment]
dont just decide to rename flags whenever you like - these flags are sent to client and mean something there!

**Comment by runningnak3d** [comment]
If that is the name in the client then of course it will stay named the way it is. Dude, me thinks you need a Xanax :)

**Comment by Shauren** [comment]
What if it does mean GO_FLAG_INTERACT_COND? (without the name having come from some debug string or anything like that)

**Comment by Rushor** [comment]
maybe we could add a costumflag for this instead (if we really don't find any relation between the scriptbehaviour and the gameobject itself - but then again that would be really cheaty again, but we also use such stuff for spellcostumattributes and creaturetemplatecostumflagextras). with a costumflag we must not touch the GO_FLAG_INTERACT_COND :)

**Comment by Shauren** [comment]
or maybe inspect values of gameobject_template? we don't use/handle many of them

**Comment by Shauren** [comment]
@runningnak3d I just finished researching gameobject flags in the client and what they do
GO_FLAG_INTERACT_COND is named correctly. It is checked by `CGGameObject_C_TypeBase::CanUse` (client) and the following condition is applied
```
if ((flags & GO_FLAG_INTERACT_COND) && !(dynamicflags & GO_DYNFLAG_LO_ACTIVATE))
    return false; // false = can't use
```

**Comment by runningnak3d** [comment]
I am not going to argue about this anymore. I am right, and you are wrong -- simple as that. Believe me, if someone hadn't stolen my project from me, this and a lot of other pull requests that are sitting here would already be in the repo.

**Comment by runningnak3d** [comment]
PS - who the fuck broke party loot? Do you guys even check shit anymore?

**Comment by jackpoz** [comment]
> who the fuck broke party loot?

maybe it was a merged PR

**Comment by Shauren** [comment]
Closing, as this change is based on incorrect assumption about GO flag meaning

**Comment by jackpoz** [comment]
about https://github.com/TrinityCore/TrinityCore/pull/17256#issuecomment-222915785 and https://github.com/TrinityCore/TrinityCore/issues/12842#issuecomment-222887209 , is there any case where a GameObject summoned like this should be looted by the same Player multiple times ? If no then that's the missing think in addition to m_respawnTime. How did it work on retail ?

**Comment by Shauren** [comment]
There are objects like the one described in #12842 which are lootable by any player from a single spawn (quests require 1x to complete so looting it completes the objective -> no longer clickable)
And then there are objects summoned by doing some action which are lootable only once, not once per player, just once. (typically used to make the world interaction more fancy)

**Comment by runningnak3d** [comment]
@Shauren in your second example, that is handled by the loot system. If there is only one item in the loot table, and it is looted -- no more loot. That applies to spell summoned GOs or event summoned GOs, or ANY GO that has loot, and it isn't party loot. As for the meaning of GO_FLAG_INTERACT_COND

From your post:
[code]
if ((flags & GO_FLAG_INTERACT_COND) && !(dynamicflags & GO_DYNFLAG_LO_ACTIVATE))
    return false; // false = can't use
[/code]

It is real easy to extrapolate the meaning. If a GO is active, but the player can not loot it (GO_DYNFLAG_LO_ACTIVATE is set when a player can loot), then they can't interact with it. No gear icon.

I really don't see how this is so hard to understand. I may not be a great programmer, but I have been reverse engineering WoW, and figuring out how the eff blizz thinks / does stuff for a LONG time.

**Pull request closed without merging**